activerecord/lib/active_record/autosave_association.rb

```diff
@@ -390,11 +390,8 @@ def save_has_one_association(reflection)
           elsif autosave != false
             key = reflection.options[:primary_key] ? send(reflection.options[:primary_key]) : id
 
-            if (autosave && record.changed_for_autosave?) || new_record? || record_changed?(reflection, record, key)
-              unless reflection.through_reflection
-                record[reflection.foreign_key] = key
-              end
-
+            if (autosave && record.changed_for_autosave?) || (!reflection.through_reflection && record_changed?(reflection, record, key))
```

Comment:
Agreed @laurocaetano. I also took `new_record?` out of `changed_for_autosave?` to be consistent and adjusted the one place where it appears to have mattered. Unsure how to abstract further without losing clarity. Please review.